Adding cloud scripting

danmcp opened the pull request:
Adding scripting to help with cloud instance creation, management and the development and testing of instructlab against cloud instances.  Currently allows for ec2 and ibm instance creation, stop, start, terminate, and ssh (including running commands).  As well as setting up a development environment with instructlab code, prereq packages, and the ability to sync changes to the repo on the remote instance.

Usage:
```bash
./cloud-instance.sh <cloud-type> <command> [options]
```

Example:

```bash
# Launch a new instance with your `EC2_INSTANCE_NAME` and `EC2_INSTANCE_TYPE` specified
# in config with the instance type including Nvidia GPU(s)
./cloud-instance.sh ec2 launch
# Clone instructlab onto the instance and setup the development environment
./cloud-instance.sh ec2 setup-rh-devenv
# Install nvidia drivers and reboot
./cloud-instance.sh ec2 install-rh-nvidia-drivers
./cloud-instance.sh ec2 ssh sudo reboot
# Install instructlab
./cloud-instance.sh ec2 pip-install-with-nvidia
# ssh to the instance
./cloud-instance.sh ec2 ssh
# Run commands on the instance through ssh and return the result
./cloud-instance.sh ec2 ssh ls -la
./cloud-instance.sh ec2 ssh "source instructlab/venv/bin/activate && ilab sysinfo"
# Sync your local git repo to the repo on the instance
./cloud-instance.sh ec2 sync
# Make changes in your local git without committing
# Sync your changes to the remote instance with a temporary commit
./cloud-instance.sh ec2 sync -c
# When you're done, stop the instance
./cloud-instance.sh ec2 stop
# While stopped and not yet pruned from your account, you can restart it if needed
./cloud-instance.sh ec2 start
# When you are done, you can delete the instance
./cloud-instance.sh ec2 terminate
```

russellb commented on scripts/cloud/cloud-instance:
```diff
@@ -0,0 +1,156 @@
+#!/bin/bash
+
+CONFIG_LOCATION=$HOME/instructlab/cloud-config
```
This will let you override the location by setting the `CONFIG_LOCATION` env var.

```suggestion
CONFIG_LOCATION=$(CONFIG_LOCATION:-$HOME/instructlab/cloud-config}
```

danmcp replied:
Fixed, thanks!

leseb commented on scripts/infra/cloud-instance:
```diff
@@ -0,0 +1,191 @@
+
+
+    echo "Waiting for instance to be in the 'running' state..."
+    while state=$(aws ec2 describe-instances \
```
Can we timeout this loop to avoid running forever?

danmcp replied:
Thanks for the suggestion!

leseb commented on scripts/infra/cloud-instance:
```diff
@@ -0,0 +1,191 @@
+        sleep 5
+    done
+
+    echo "Instance is now running."
```
Print instance ID?

danmcp replied:
Part of the intention of this utility is to not have to deal with the instance id.  Printing it out might make the user think it was needed for something.

danmcp replied:
Changed it to print the name and id next to it in ()

leseb reviewed (commented):
I'm failing to understand why this helper script is needed in this repository? Can you help me? Thanks!

danmcp commented:
> I'm failing to understand why this helper script is needed in this repository? Can you help me? Thanks!

@leseb Perhaps there is a better location.  But the intention is to create a set of useful tooling to make it easy to script the launching of an instance and run instructlab.

leseb commented:
> > I'm failing to understand why this helper script is needed in this repository? Can you help me? Thanks!
> 
> @leseb Perhaps there is a better location. But the intention is to create a set of useful tooling to make it easy to script the launching of an instance and run instructlab.

I see the potential usefulness of this script for users with AWS accounts looking to experiment. However, I have some concerns about adding it to this repository. Accepting it might open the door to contributions for other cloud providers and local virtualized environments, which could create maintenance challenges. Everyone's workflow is unique, and it will be difficult to accommodate all variations. I'm also worried about setting a precedent without clear limits on what we accept. Additionally, the script appears to be compatible only with Fedora, which could cause issues for users on Debian-based distributions.

I'd like to hear what other maintainers think about this too. Thanks!

danmcp commented:
> I see the potential usefulness of this script for users with AWS accounts looking to experiment. However, I have some concerns about adding it to this repository. Accepting it might open the door to contributions for other cloud providers and local virtualized environments, which could create maintenance challenges. Everyone's workflow is unique, and it will be difficult to accommodate all variations. I'm also worried about setting a precedent without clear limits on what we accept. Additionally, the script appears to be compatible only with Fedora, which could cause issues for users on Debian-based distributions.

The script does call out it's developed for Red Hat based systems in a way that would allow someone to add another distro later.  Note that as of today, the InstructLab READMEs also don't support anything other than Red Hat based linux.

As far as clear limits, this isn't core InstructLab code and could be considered best effort and only supporting an "easy path".

I don't have a strong opinion that this work must go into instructlab.  The reason I have started here is it seems like it could be helpful for others working on the project.

russellb commented:
> > > I'm failing to understand why this helper script is needed in this repository? Can you help me? Thanks!
> > 
> > 
> > @leseb Perhaps there is a better location. But the intention is to create a set of useful tooling to make it easy to script the launching of an instance and run instructlab.
> 
> I see the potential usefulness of this script for users with AWS accounts looking to experiment. However, I have some concerns about adding it to this repository. Accepting it might open the door to contributions for other cloud providers and local virtualized environments, which could create maintenance challenges. Everyone's workflow is unique, and it will be difficult to accommodate all variations. I'm also worried about setting a precedent without clear limits on what we accept. Additionally, the script appears to be compatible only with Fedora, which could cause issues for users on Debian-based distributions.
> 
> I'd like to hear what other maintainers think about this too. Thanks!

I'm happy having it here. Nobody is required to use it, but it's a way to collaborate for those that use it.  If it gets out of hand, we can always move it later.  I don't have any better ideas for a place to share dev env scripts right now.

I don't think it should have to support all possible target dev environments out of the gate to go in. It just needs to be helpful to multiple people.

nathan-weinberg commented:
Just to chime in, I'm fine with these scripts living here as well.

danmcp commented:
Noting that I've got everything to a functional starting point for ec2 and ibm and the code is ready for any additional reviews/feedback.

danmcp commented on scripts/infra/cloud-instance:
```diff
@@ -0,0 +1,389 @@
+    local cloud_type=$1
+    $BASH_SOURCE $cloud_type ssh -n $INSTANCE_NAME sudo dnf update -y
+    $BASH_SOURCE $cloud_type ssh -n $INSTANCE_NAME sudo dnf install git gcc make pip python3 python3-devel -y
+    $BASH_SOURCE $cloud_type ssh -n $INSTANCE_NAME sudo dnf install g++ -y || \
```
This will work but will need to be in quotes so the || applies to the ssh.

leseb replied:
Good point, I overlooked that. Including everything into one line makes the fallback easier to read :)

danmcp commented on scripts/infra/cloud-instance:
```diff
@@ -0,0 +1,389 @@
+    local branch="$(git symbolic-ref HEAD 2>/dev/null)"
+    branch=${branch##refs/heads/}
+    SCRIPT_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
+    if [ "$temp_commit" = true ] && [ ! -z "$(git status --porcelain=v1 2>/dev/null)" ]; then
```
Thanks!

danmcp commented on scripts/infra/cloud-instance:
```diff
@@ -0,0 +1,389 @@
+MAX_ITERS=12
+INSTRUCTLAB_CLOUD_CONFIG=${INSTRUCTLAB_CLOUD_CONFIG:-$HOME/.instructlab/cloud-config}
+
+if test -f $INSTRUCTLAB_CLOUD_CONFIG ; then
```
Fixed. Thanks!

danmcp commented on scripts/infra/cloud-instance:
```diff
@@ -0,0 +1,389 @@
+
+ec2() {
+    local cmdname=$1; shift
+    if type "ec2__$cmdname" >/dev/null 2>&1; then
```
I've changed this to use -t to avoid finding anything other than a function.

danmcp commented on scripts/infra/cloud-instance:
```diff
@@ -0,0 +1,389 @@
+
+ibm() {
+    local cmdname=$1; shift
+    if type "ibm__$cmdname" >/dev/null 2>&1; then
```
Same resolution as above.

leseb reviewed (commented):
I'd like to hold this until https://github.com/instructlab/instructlab/pull/1211 is merged since it enables the shellcheck linter. Once merged, it's +1 for me since other maintainers agreed to have the script.

Thanks for addressing all the comments, could you in PR description add a workflow example of the script usage? So we can see a workflow example of using the script? Thanks!

danmcp commented:
> I'd like to hold this until #1211 is merged since it enables the shellcheck linter. Once merged, it's +1 for me since other maintainers agreed to have the script.

Thanks for the review!

> 
> Thanks for addressing all the comments, could you in PR description add a workflow example of the script usage? So we can see a workflow example of using the script? Thanks!

Done!

leseb commented:
@Mergifyio rebase

danmcp commented:
> Can you rename `scripts/infra/cloud-instance` to `scripts/infra/cloud-instance.sh` so that the CI can pick it up? Thanks!

Done

leseb commented on scripts/infra/instructlab-cloud-config-example:
```diff
@@ -0,0 +1,18 @@
+EC2_INSTANCE_TYPE="m4.large"
```
The specs look relatively small depending on what we want to do. Can we add a section to guide users on choosing the optimal cloud instance type for their usage? It looks m4.large does not have any GPU, we could recommend a type that can take advantage of that?

danmcp replied:
I added additional instance type recommendations for EC2 and IBM

leseb commented on scripts/infra/cloud-instance.sh:
```diff
@@ -0,0 +1,399 @@
+#!/bin/bash -e
```
```suggestion
# SPDX-License-Identifier: Apache-2.0

#!/usr/bin/env -S bash -e
```

danmcp replied:
Note the shebang has to be the first line.

danmcp commented on scripts/infra/README.md:
```diff
@@ -0,0 +1,54 @@
+
+## Prereqs
+
+You'll need to setup the respective cloud cli before using cloud-instance.sh
```
Done

danmcp commented on scripts/infra/README.md:
```diff
@@ -0,0 +1,54 @@
+
+### IBM
+
+- ibmcloud (the IBM Cloud cli)
```
Done

danmcp commented on scripts/infra/README.md:
```diff
@@ -0,0 +1,54 @@
+
+## Config
+
+You'll need to specify the cloud-instance.sh config before getting started. [instructlab-cloud-config-example](instructlab-cloud-config-example) is provided as a template. By default, you'll need to place your config in ~/.instructlab/cloud-config, or you can specify the location as the environment variable `INSTRUCTLAB_CLOUD_CONFIG`
```
Done

danmcp commented on scripts/infra/README.md:
```diff
@@ -0,0 +1,54 @@
+
+## Config
+
+You'll need to specify the cloud-instance.sh config before getting started. [instructlab-cloud-config-example](instructlab-cloud-config-example) is provided as a template. By default, you'll need to place your config in ~/.instructlab/cloud-config, or you can specify the location as the environment variable `INSTRUCTLAB_CLOUD_CONFIG`
```
Examples provided.

nathan-weinberg commented on scripts/infra/README.md:
```diff
@@ -0,0 +1,71 @@
+
+## Prereqs
+
+You'll need to setup the respective cloud cli before using [cloud-instance.sh](cloud-instance.sh)
```
```suggestion
You'll need to setup the respective cloud CLI before using [cloud-instance.sh](cloud-instance.sh)
```

danmcp replied:
Fixed, thanks!

nathan-weinberg commented on scripts/infra/README.md:
```diff
@@ -0,0 +1,71 @@
+
+- [ibmcloud](https://cloud.ibm.com/docs/cli) (the IBM Cloud CLI)
+- An sshkey created within IBM Cloud. Config defaults to `whoami` for key name.
+- A floating ip created within IBM Cloud. Config defaults to `whoami` for floating ip name.
```
```suggestion
- A floating IP created within IBM Cloud. Config defaults to `whoami` for floating IP name.
```

danmcp replied:
Fixed, thanks!

nathan-weinberg commented on scripts/infra/README.md:
```diff
@@ -0,0 +1,71 @@
+
+## Config
+
+You'll need to specify the cloud-instance.sh config before getting started. [instructlab-cloud-config-example](instructlab-cloud-config-example) is provided as a template. By default, you'll need to place your config in `~/.instructlab/cloud-config`, or you can specify the location as the environment variable `INSTRUCTLAB_CLOUD_CONFIG`
```
```suggestion
You'll need to specify the `cloud-instance.sh` config before getting started. [instructlab-cloud-config-example](instructlab-cloud-config-example) is provided as a template. By default, you'll need to place your config in `~/.instructlab/cloud-config`, or you can specify the location as the environment variable `INSTRUCTLAB_CLOUD_CONFIG`
```

danmcp replied:
Fixed, thanks!

nathan-weinberg commented on scripts/infra/README.md:
```diff
@@ -0,0 +1,71 @@
+
+## Usage
+
+`./cloud-instance.sh <cloud-type)> <command> [options]`
```
```suggestion
`./cloud-instance.sh <cloud-type> <command> [options]`
```

leseb reviewed (approved):
I could add more comments in the README, but you've already addressed a lot, and I'm satisfied with the current state. I don't want to make it seem like I'm being overly critical of your contribution. We can always update the README.md again if needed. Thank you for your patience and great work!

leseb commented:
@Mergifyio rebase

leseb commented:
Mergify is not merging because of https://github.com/instructlab/instructlab/pull/1296

russellb commented on scripts/infra/README.md:
```diff
@@ -0,0 +1,71 @@
+./cloud-instance.sh ec2 sync
+# Make changes in your local git without committing
+# Sync your changes to the remote instance with a temporary commit
+./cloud-instance.sh ec2 sync -c
```
```suggestion
./cloud-instance.sh ec2 sync -c
# When you're done, stop your instance
./cloud-instanfce.sh ec2 stop
# While stopped and not yet pruned from your account, you can restart it if needed
./cloud-instance.sh ec2 start
```

danmcp replied:
Suggestion accepted and I added terminate as well.  Thanks!

russellb reviewed (commented):
This is great! I used it for my entire dev workflow on this PR: https://github.com/instructlab/instructlab/pull/1297

It worked perfectly!

I only have very minor suggestions, and then I'd like to merge it. Thank you!

russellb commented on scripts/infra/README.md:
```diff
@@ -0,0 +1,71 @@
+./cloud-instance.sh ec2 install-rh-nvidia-drivers
+./cloud-instance.sh ec2 ssh sudo reboot
+# Install instructlab
+./cloud-instance.sh ec2 pip-install-with-nvidia
```
to be clear, I don't think we need to block on this before merging, because the code is doing what our docs say ... can sort it out later

leseb commented:
Putting on hold so that @russellb can be addressed since this is going to be merged once https://github.com/instructlab/instructlab/pull/1296 is merged.

alimaredia commented on scripts/infra/cloud-instance.sh:
```diff
@@ -0,0 +1,400 @@
+    "${BASH_SOURCE[0]}" "$cloud_type" ssh -n "$INSTANCE_NAME" sudo dnf config-manager --add-repo https://developer.download.nvidia.com/compute/cuda/repos/rhel9/x86_64/cuda-rhel9.repo
+    "${BASH_SOURCE[0]}" "$cloud_type" ssh -n "$INSTANCE_NAME" sudo dnf clean expire-cache
+    "${BASH_SOURCE[0]}" "$cloud_type" ssh -n "$INSTANCE_NAME" sudo dnf module install nvidia-driver:latest-dkms -y
+    "${BASH_SOURCE[0]}" "$cloud_type" ssh -n "$INSTANCE_NAME" sudo dnf install cuda-toolkit nvidia-gds -y
```
FYI this is a more comprehensive list of packages I installed to get vllm working as a serving backend. 
```shell
sudo dnf install -y cuda-toolkit nvidia-gds libnccl-devel libcudnn8 nvidia-driver-NVML nvidia-driver-cuda-libs`
```

I can open up another PR with this if you'd rather not include it in this one.

danmcp replied:
I went ahead and included.  Thanks!

danmcp commented on scripts/infra/instructlab-cloud-config-example:
```diff
@@ -0,0 +1,28 @@
+# Nvidia 4x A10 GPU, 96GB VRAM
+#EC2_INSTANCE_TYPE="g5.12xlarge"
+# Nvidia 1x A10 GPU, 24GB VRAM
+#EC2_INSTANCE_TYPE=g5.4xlarge"
```
Fixed, thanks!

danmcp commented on scripts/infra/README.md:
```diff
@@ -0,0 +1,71 @@
+# in config with the instance type including Nvidia GPU(s)
+./cloud-instance.sh ec2 launch
+# Clone instructlab onto the instance and setup the development environment
+./cloud-instance.sh ec2 setup-devenv
```
Fixed, thanks!

danmcp commented on scripts/infra/cloud-instance.sh:
```diff
@@ -0,0 +1,400 @@
+
+pip_install_with_nvidia() {
+    local cloud_type=$1
+    "${BASH_SOURCE[0]}" "$cloud_type" ssh -n "$INSTANCE_NAME" "pushd instructlab && source venv/bin/activate && pip cache remove llama_cpp_python && pip install . -C cmake.args='-DLLAMA_CUBLAS=on' && ilab"
```
I like the suggestion.  Going with it.  Changed for AMD as well.

danmcp commented:
> This is great! I used it for my entire dev workflow on this PR: #1297
> 
> It worked perfectly!
> 
> I only have very minor suggestions, and then I'd like to merge it. Thank you!

@russellb Thanks for trying it out!

russellb reviewed (approved):
thank you!

leseb reviewed (approved):
Thanks!

russellb commented:
> @russellb @leseb seems the automerging isn't happening due to the below ![image](https://private-user-images.githubusercontent.com/31703736/338610607-4667dabf-92ce-4331-bf7e-5990f7838a6a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgxMzU5NzIsIm5iZiI6MTcxODEzNTY3MiwicGF0aCI6Ii8zMTcwMzczNi8zMzg2MTA2MDctNDY2N2RhYmYtOTJjZS00MzMxLWJmN2UtNTk5MGY3ODM4YTZhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjExVDE5NTQzMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRiYzFlOGQzY2U5ZDc4NzMyNThmYTA5YWFkZGIxZjU2MGQxOWQ5ZWFmZmMwN2VhODJhMjM5YmFlMTVkNTAyMGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.ln3Nw9s44v49iCz1dF1urRgK1xeIXHvDVkk_YHJWNng)

It's not merging because the lint job didn't run. I'll try to fix this.

russellb commented:
This will fix the auto-merge issue in the future: https://github.com/instructlab/instructlab/pull/1330

russellb commented:
Since the auto-merge config was the only thing holding this up, I'm just going to manually merge it. Thanks, Dan!
